pkgs/build-support/deno/buildDenoApplication.nix

```diff
@@ -0,0 +1,474 @@
+{
+  lib,
+  stdenvNoCC,
+  fetchDenoDeps,
+  unzip,
+  xxd,
+  util-linux,
+  glibc,
+  stdenv,
+  fetchurl,
+  deno,
+  jq,
+}@topLevelArgs:
+{
+  # Name of the package
+  name ? "${args.pname}-${args.version}",
+  # Use this deno binary
+  deno ? topLevelArgs.deno,
+
+  # Name for the vendored dependencies tarball
+  denoDepsName ? "${name}-deno-deps",
+  # The vendored dependencies. The output structure of that derivation is still tbd
+  denoDeps ? (
+    fetchDenoDeps {
+      inherit
+        src
+        srcs
+        sourceRoot
+        prePatch
+        patches
+        postPatch
+        ;
+      name = denoDepsName;
+      outputHash = denoDepsHash;
+    }
+  ),
+  # Hash of the vendored dependencies. Leave empty if you dont know it yet
+  denoDepsHash ? "",
+
+  # Configuration for finding the main script and the config file 
+  # The main script you would run with `deno run`
+  mainScript,
+  # The deno.json config file. 
+  # TODO: Add support for deno.jsonc and allow for projects without deno.json
+  denoJson ? "",
+  # Customize the deno lockfile location
+  denoLock ? "",
+  # Deno runtime flags that will be passed to deno during execution. Like `--allow-net` or `-A`
+  #
+  # Example: ["--allow-net", "--allow-read"] or ["-A"]
+  runtimeFlags ? [ ],
+  # TODO: Allow projects with custom vendored dependencies
+
+  # Configuration for how the application should be run
+  # Create a temporary directory, where deno can cache some things. Stops deno from whining about not being able to write to the deno directory.
+  # Not used if compile is true.
+  writableDenoDirectory ? true,
+  # Just produce a single binary, instead of a whole directory.
+  compile ? false,
+
+  # Idk if I actually need to specify these explitly
+  # Both npm and rust explicity specify src, srcs, and sourceRoot
+  src ? null,
+  srcs ? null,
+  sourceRoot ? null,
+  nativeBuildInputs ? [ ],
+  passthru ? { },
+  patches ? [ ],
+  meta ? { },
+  prePatch ? "",
+  postPatch ? "",
+  buildInputs ? [ ],
+  preUnpack ? null,
+  unpackPhase ? null,
+  postUnpack ? null,
+  # We want parallel builds by default
+  # TODO: idk if deno even supports something like parallel building
+  enableParallelBuilding ? false,
+
+  #   src,
+  #   script,
+  #   depsHash ? "",
+  #   denoJson ? "deno.json",
+  #   # Create a temporary directory, where deno can cache some things. Stops deno from whining about not being able to write to the deno directory.
+  #   # Not used if compile is true.
+  #   writableDenoDirectory ? true,
+  #   # Just produce a single binary, instead of a whole directory.
+  #   compile ? false,
+  #   ...
+  # }@args:
+
+  # References used:
+  # pkgs/build-support/rust/build-rust-package/default.nix
+  # pkgs/build-support/node/build-npm-package/default.nix
+  # pkgs/build-support/go/module.nix
+  # 
+  # Looked at and didnt understand:
+  # pkgs/development/compilers/nim/build-nim-package.nix
```

Comment:
`build-nim-package.nix` is convoluted and hard to understand because it uses fixed-point style right off, it does lockfile processing, and it has a function for overriding the lockfile.

The solution might be too much of a hack to be a good example but think the Nim problem is conceptually similar in that the Nim package description format and lockfiles are not reproducible by Nim standards. The current Nim solution is to use a tool to create a Nix specific lockfile when manually updating packages and commit the lockfile into Nixpkgs. There isn't anything so show that this will scale well into the future.

The solution that I've been working on is to [reuse](https://git.sr.ht/~ehmry/nim_lk#nimble-to-cyclonedx-sbom) an independent software-bill-of-materials standard and add some Nix annotations but nothing has been merged to Nixpkgs. This comes with an alternative implementation of `build-nim-package.nix` - [build-nim-sbom.nix](https://git.sr.ht/~ehmry/nim_lk/tree/trunk/build-nim-sbom.nix).